In [1]:
# If you havent installed pandas and SQLite3
'''
pip install pandas
pip install sqlite3
'''

'\npip install pandas\npip install sqlite3\n'

In [2]:
import pandas as pd
import os
import sqlite3
#please change your working directory according to the file in your machine
os.chdir(r'/Users/mint/Code') #replace /Users/mint/Code

# Scen 2 Task 1

In [3]:
# Loading excel sheets into dataframe
df1 = pd.read_excel('Genie Dataset 1.xlsx', sheet_name=0)
#df1a = pd.read_excel('Genie Dataset 1.xlsx', sheet_name=1)

df2 = pd.read_excel('Genie Dataset 2.xlsx', sheet_name=0)
#df2a = pd.read_excel('Genie Dataset 1.xlsx', sheet_name=1)

df3 = pd.read_excel('Genie Dataset 3.xlsx', sheet_name=0)
#df2a = pd.read_excel('Genie Dataset 1.xlsx', sheet_name=1)

df4 = pd.read_excel('Genie Dataset 4.xlsx', sheet_name=0)
#df3a = pd.read_excel('Genie Dataset 1.xlsx', sheet_name=1)


In [4]:
def df1_info(df1):
    print("------------ Shape ------------")
    print(df1.shape)
    print("------------ Info ------------")
    print(df1.info())
    print("------------ Columns ------------")
    print(df1.columns)
    print("------------ Quantiles ------------")
    print(df1.describe())
    print("------------ NA ------------")
    print(df1.isnull().sum())
print(df1_info(df1))

------------ Shape ------------
(277, 23)
------------ Info ------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date Settled   277 non-null    datetime64[ns]
 1   Agrt No.       277 non-null    object        
 2   Reg No.        277 non-null    object        
 3   Agrt Date      277 non-null    datetime64[ns]
 4   Block No       0 non-null      float64       
 5   Loan Amt       277 non-null    int64         
 6   Term Charges   277 non-null    int64         
 7   Total Payable  277 non-null    int64         
 8   Principle Amt  277 non-null    int64         
 9   Rebate         277 non-null    int64         
 10  GL Bal         277 non-null    int64         
 11  Last Pd Amt    277 non-null    int64         
 12  Tot Instal Pd  277 non-null    int64         
 13  O/D Int        277 non-null    int64         
 14  P

In [5]:
def df3_info(df3):
    print("------------ Shape ------------")
    print(df3.shape)
    print("------------ Info ------------")
    print(df3.info())
    print("------------ Columns ------------")
    print(df3.columns)
    print("------------ Quantiles ------------")
    print(df3.describe())
    print("------------ NA ------------")
    print(df3.isnull().sum())
print(df3_info(df3))

------------ Shape ------------
(554, 53)
------------ Info ------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 554 entries, 0 to 553
Data columns (total 53 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Branch              554 non-null    object        
 1   Agrt No.            554 non-null    object        
 2   Hirer IC            554 non-null    object        
 3   Age                 554 non-null    int64         
 4   State               554 non-null    object        
 5   Occupation          11 non-null     object        
 6   Marital             554 non-null    object        
 7   Race                554 non-null    object        
 8   Gender              554 non-null    object        
 9   Total Income        554 non-null    int64         
 10  Dealer ID           553 non-null    object        
 11  Reg No.             554 non-null    object        
 12  Car Made            554 non-null 

In [6]:
def df4_info(df4):
    print("------------ Shape ------------")
    print(df4.shape)
    print("------------ Info ------------")
    print(df4.info())
    print("------------ Columns ------------")
    print(df4.columns)
    print("------------ Quantiles ------------")
    print(df4.describe())
    print("------------ NA ------------")
    print(df4.isnull().sum())
print(df4_info(df4))

------------ Shape ------------
(602, 16)
------------ Info ------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 602 entries, 0 to 601
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Agrt No         602 non-null    object        
 1   Agrt Date       602 non-null    datetime64[ns]
 2   IC No.          602 non-null    object        
 3   Dealer ID       602 non-null    object        
 4   Reg No          602 non-null    object        
 5   Vehicle         602 non-null    object        
 6   Loan Amt        602 non-null    int64         
 7   Mthly Instal    602 non-null    int64         
 8   Int Rate        602 non-null    float64       
 9   Tenor           602 non-null    int64         
 10  GL              602 non-null    int64         
 11  Last Paid Date  602 non-null    datetime64[ns]
 12  Last Paid Amt   602 non-null    int64         
 13  Arrears         602 non-null    int64

In [7]:
print(df1['Agrt No.'].unique())
# print(df3['Agrt No.'].unique())
# print(df4['Agrt No'].unique())

['A0000001' 'A0000002' 'A0000003' 'A0000004' 'A0000005' 'A0000006'
 'A0000007' 'A0000008' 'A0000009' 'A0000010' 'A0000011' 'A0000012'
 'A0000013' 'A0000014' 'A0000015' 'A0000016' 'A0000017' 'A0000018'
 'A0000019' 'A0000020' 'A0000021' 'A0000022' 'A0000023' 'A0000024'
 'A0000025' 'A0000026' 'A0000027' 'A0000028' 'A0000029' 'A0000030'
 'A0000031' 'A0000032' 'A0000033' 'A0000034' 'A0000035' 'A0000036'
 'A0000037' 'A0000038' 'A0000039' 'A0000040' 'A0000041' 'A0000042'
 'A0000043' 'A0000044' 'A0000045' 'A0000046' 'A0000047' 'A0000048'
 'A0000049' 'A0000050' 'A0000051' 'A0000052' 'A0000053' 'A0000054'
 'A0000055' 'A0000056' 'A0000057' 'A0000058' 'A0000059' 'A0000060'
 'A0000061' 'A0000062' 'A0000063' 'A0000064' 'A0000065' 'A0000066'
 'A0000067' 'A0000068' 'A0000069' 'A0000070' 'A0000071' 'A0000072'
 'A0000073' 'A0000074' 'A0000075' 'A0000076' 'A0000077' 'A0000078'
 'A0000079' 'A0000080' 'A0000081' 'A0000082' 'A0000083' 'A0000084'
 'A0000085' 'A0000086' 'A0000087' 'A0000088' 'A0000089' 'A0000

### Findings
Based on data exploration, "Agrt No." and "Agrt No" can be used as the common key to index and combine data.
Since the files are not too big, we will be combining all sheets into one dataset before loading into SQL. 

Since it is unclear what the data will be used for, I will merge the data sets without removing any rows

In [8]:
print(df1['Agrt No.'].dtype)
print(df3['Agrt No.'].dtype)
print(df4['Agrt No'].dtype)

object
object
object


In [9]:
# Merge data frames 
# Merge on df1 and df3 first due to column names 'Agrt No.'
merged_df = pd.merge(df1, df3, left_on='Agrt No.', right_on='Agrt No.', how='outer')
# Merge on df4
final_merged_df = pd.merge(merged_df, df4, left_on='Agrt No.', right_on='Agrt No', how='outer')
print(final_merged_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 879 entries, 0 to 878
Data columns (total 91 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Date Settled        277 non-null    datetime64[ns]
 1   Agrt No.            831 non-null    object        
 2   Reg No._x           277 non-null    object        
 3   Agrt Date_x         277 non-null    datetime64[ns]
 4   Block No            0 non-null      float64       
 5   Loan Amt_x          277 non-null    float64       
 6   Term Charges        277 non-null    float64       
 7   Total Payable       277 non-null    float64       
 8   Principle Amt       277 non-null    float64       
 9   Rebate_x            277 non-null    float64       
 10  GL Bal_x            277 non-null    float64       
 11  Last Pd Amt_x       277 non-null    float64       
 12  Tot Instal Pd       277 non-null    float64       
 13  O/D Int             277 non-null    float64       

In [10]:
# Connect to SQlite database/create database
# this data base will be created in your os path
conn = sqlite3.connect('genie_database.db')
# Create cursor object to execute SQL queries
cursor = conn.cursor()

In [11]:
# Load merged df into db
final_merged_df.to_sql('final_table', conn, if_exists='replace', index=False)

879

In [12]:
# Query the database to check the first few rows
cursor.execute("SELECT * FROM final_table LIMIT 5")
rows = cursor.fetchall()

# Print the retrieved rows
for row in rows:
    print(row)

('2024-05-06 00:00:00', 'A0000001', 'WRJ****', '2022-01-27 00:00:00', None, 18000.0, 9000.0, 27000.0, 11640.0, 2600.0, 0.0, 24600.0, 24400.0, 8.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 'D0000001', 'CONVENTIONAL', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None)
('2023-04-05 00:00:00', 'A0000002', 'WTU****', '2022-02-08 00:00:00', None, 24000.0, 6840.0, 30840.0, 17960.0, 2600.0, 0.0, 18040.0, 28240.0, 8.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 'D0000002', 'CONVENTIONAL', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None

In [13]:
# Commit changes and close the connection
conn.commit()
conn.close()

# Scen 2 Task 2 
Generate HP ageing buckets

Task 2: Generate hp aging bucket by number of account and general ledger balance snapshot based on Genie Dataset 1 and Genie Dataset 4. 
Aging buckets refer to category M0 : 0 days, category M1 : 1 – 30 days, category M2 : 31– 60 days, category M3 : 61 – 90 days, category > M3 : > 90 days. 
Aging is defined based on days different between reporting day and contract commencement date. Candidate could provide the result in python format or using any BI Tools.

HP refers to hire purchase: difference between reporting date and contract commencement date [Agrt Date, Date Settled] and [Agrt Date, Last paid date]
General Ledger: GL or GL Bal

In [14]:
print(df1.info())
print(df1.head)

print(df4.info())
print(df4.head)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 277 entries, 0 to 276
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   Date Settled   277 non-null    datetime64[ns]
 1   Agrt No.       277 non-null    object        
 2   Reg No.        277 non-null    object        
 3   Agrt Date      277 non-null    datetime64[ns]
 4   Block No       0 non-null      float64       
 5   Loan Amt       277 non-null    int64         
 6   Term Charges   277 non-null    int64         
 7   Total Payable  277 non-null    int64         
 8   Principle Amt  277 non-null    int64         
 9   Rebate         277 non-null    int64         
 10  GL Bal         277 non-null    int64         
 11  Last Pd Amt    277 non-null    int64         
 12  Tot Instal Pd  277 non-null    int64         
 13  O/D Int        277 non-null    int64         
 14  Postage        277 non-null    float64       
 15  Misc           277 non-

In [15]:
# merge df1 and df4
task2_df = pd.merge(df1, df4, left_on='Agrt No.', right_on='Agrt No', how='outer')

In [16]:
print(task2_df.columns)

Index(['Date Settled', 'Agrt No.', 'Reg No.', 'Agrt Date_x', 'Block No',
       'Loan Amt_x', 'Term Charges', 'Total Payable', 'Principle Amt',
       'Rebate_x', 'GL Bal', 'Last Pd Amt', 'Tot Instal Pd', 'O/D Int',
       'Postage', 'Misc', 'Repo Exp', 'Storage', 'Legal Exp', 'LOD', 'JPJ Exp',
       'Dealer ID_x', 'Loan Type', 'Agrt No', 'Agrt Date_y', 'IC No.',
       'Dealer ID_y', 'Reg No', 'Vehicle', 'Loan Amt_y', 'Mthly Instal',
       'Int Rate', 'Tenor', 'GL ', 'Last Paid Date', 'Last Paid Amt',
       'Arrears', 'Mth Due', 'Rebate_y'],
      dtype='object')


In [17]:
# difference in days create column called aging
# for df1 columns
task2_df['Ageing'] = (task2_df['Date Settled'] - task2_df['Agrt Date_x']).dt.days
# for df2 columns
task2_df['Ageing'] = (task2_df['Last Paid Date'] - task2_df['Agrt Date_y']).dt.days

In [18]:
print(task2_df.head)

<bound method NDFrame.head of     Date Settled  Agrt No.  Reg No. Agrt Date_x  Block No  Loan Amt_x  \
0     2024-05-06  A0000001  WRJ****  2022-01-27       NaN     18000.0   
1     2023-04-05  A0000002  WTU****  2022-02-08       NaN     24000.0   
2     2024-05-17  A0000003  WXT****  2022-02-09       NaN      8400.0   
3     2024-03-14  A0000004  WVJ****  2022-02-09       NaN     16200.0   
4     2024-03-13  A0000005  WUF****  2022-02-17       NaN      9840.0   
..           ...       ...      ...         ...       ...         ...   
874          NaT       NaN      NaN         NaT       NaN         NaN   
875          NaT       NaN      NaN         NaT       NaN         NaN   
876          NaT       NaN      NaN         NaT       NaN         NaN   
877          NaT       NaN      NaN         NaT       NaN         NaN   
878          NaT       NaN      NaN         NaT       NaN         NaN   

     Term Charges  Total Payable  Principle Amt  Rebate_x  ...  Mthly Instal  \
0          90

In [19]:
# Combine Agrt No columns, filling in missing values from one column with values from another
task2_df['Agrt_No_combined'] = task2_df['Agrt No.'].combine_first(task2_df['Agrt No'])

In [20]:
# General Ledger columns
#replace NaN values in GL and GL Bal
task2_df['GL Bal'] = task2_df['GL Bal'].fillna(0)
task2_df['GL '] = task2_df['GL '].fillna(0)
# combine different GL and GL Bal columns
task2_df['total_GL_balance'] = task2_df['GL Bal'] + task2_df['GL ']

In [21]:
# Define age buckets based on the column ageing
def ageing_catergory(days):
    if days == 0:
        return 'M0'
    elif 1 <= days <= 30:
        return 'M1'
    elif 31 <= days <= 60:
        return 'M2'
    elif 61 <= days <= 90:
        return 'M3'
    else:
        return '>M3'

In [22]:
# Creating a new column for 'Aging 
task2_df['Ageing Categories'] = task2_df['Ageing'].apply(ageing_catergory)

In [23]:
# creating table
# group by aging bucket, count unique accounts and sum the general ledger balance
ageing_summary = task2_df.groupby('Ageing Categories').agg(
    num_accounts=('Agrt_No_combined', 'nunique'),  # use 'nunique' to count unique account numbers
    total_gl_balance=('total_GL_balance', 'sum')   # sum of general ledger 
).reset_index()

print(ageing_summary)

  Ageing Categories  num_accounts  total_gl_balance
0               >M3           879        13992400.0


# Scen 2 Task 3
Based on HP OS, what is company risk profile and illustrate current account performance summary ie. Current active loan holder paid installment in term of month and installment period position.

Considerations of risk profile: 
- loan holders installment period
- whether they are paying on time

In [24]:
performance_summary = task2_df.groupby('Ageing Categories').agg(
    num_accounts=('Agrt_No_combined', 'nunique'),          
    total_loan_amount=('Loan Amt_x', 'sum'),               
    total_gl_balance=('total_GL_balance', 'sum'),          
    avg_remaining_months=('Mth Due', 'mean')
).reset_index()

print(performance_summary)

  Ageing Categories  num_accounts  total_loan_amount  total_gl_balance  \
0               >M3           879          6113700.0        13992400.0   

   avg_remaining_months  
0              0.935216  


## Task 3 outcome
Due to all accounts falling in the ">M3" catergory, this indicates a high probability of default from all accounts. Monitoring these accounts are critical to managing credit risk. 

However, since the average remaining months for loan due is 0.9, this indicates that the accounts are ahead of their payments. 
Total Gross Ledger is double that of the total loan amount, the companies position is still healthy. As a precautionary measure, we can estimate potential defaults for loans in the >M3 catergories based on historical recovery rates from datasaet 1 and closely monitor overdue loan accounts of those in dataset 4. 

# Scen 3
Scenario: You are providing Genie Dataset 2. Can you provide your insights and recommendations inline with the following questions 
Question 1:  What do you think about a company underwriting risk profile? Please elaborate your thought and how we could improve from current situation.


In [25]:
df2.head()

Submission Date Application No. Prepared Date Judge Status  \
0      2021-12-08      APP0000001    2022-02-24    DISBURSED   
1      2021-12-10      APP0000002    2021-12-13       CANCEL   
2      2021-12-13      APP0000003    2021-12-16       CANCEL   
3      2021-12-14      APP0000004    2021-12-20       CANCEL   
4      2021-12-14      APP0000005    2021-12-21       CANCEL   

                                       Judge Remarks  \
0  Incomplete customer details. _x000D_\n_x000D_\...   
1  APPLICANT WORKING AS A SUPERVISOR FOR 2 YEARS ...   
2  REJECTE GTOR ARMY_x000D_\n_x000D_\nHIRER NOT P...   
3  -PENDING ANOTHER 3 MTHS P/SLIP AND EPF OR BANK...   
4  PENDING BANK STATEMENT_x000D_\nAPPLICANT REQUE...   

                                     Prepare Remarks  Agrt No. Approval Date  \
0  APPLICANT IS OWN BUSINESS, OWNING 2 SOLE-PROPS...  A0000008    2022-02-24   
1                                    DONE AMENDMENT.       NaN    2021-12-30   
2  REJECTE GTOR ARMY_x000D_\n_x000D_\nHIRER NOT P...       NaN    2021-12-16   
3  -PENDING ANOTHER 3 MTHS P/SLIP AND EPF OR BANK...       NaN    2021-12-20   
4  APPLICANT REQUEST FOR PROTON X70 MANUFACTURED ...       NaN    2021-12-21   

     Hirer NRIC Car Plate No.  ... Appeal By Date Appeal  \
0  77110808****       DEF****  ...      YEAP  2022-02-23   
1  85101613****       QS6****  ...    JINFEI  2022-04-04   
2  99021311****       ALA****  ...       NaN         NaT   
3  98072814****       WA2****  ...       NaN         NaT   
4  90101109****       VDB****  ...       NaN         NaT   

   First Approved Date(N)  First Approved By(N)  Ins Date  First Reject Date  \
0                     NaN                   NaN       NaN                NaN   
1                     NaN                   NaN       NaN                NaN   
2                     NaN                   NaN       NaN                NaN   
3                     NaN                   NaN       NaN                NaN   
4                     NaN                   NaN       NaN                NaN   

   First Reject By     Loan Type  Dealer ID.1 H Yrs Svr  
0              NaN  CONVENTIONAL      D000001       0.0  
1              NaN  CONVENTIONAL      D000001       2.0  
2              NaN  CONVENTIONAL      D000001       0.0  
3              NaN  CONVENTIONAL      D000001       0.0  
4              NaN  CONVENTIONAL      D000001       0.0  

[5 rows x 74 columns]

In [26]:
# Based on the data, Judge Status identifies if loan was approved and specific details which will reveal key patterns related to loan applications and thier approval status.
status_distribution = df2['Judge Status'].value_counts()
print(status_distribution)

Judge Status
CANCEL       237
DISBURSED     16
Name: count, dtype: int64


In [27]:
# Analyze reasons for rejection by looking at Judge Remarks for rejected/canceled applications
rejected_data = df2[df2['Judge Status'].isin(['CANCEL', 'REJECT'])]
rejection_reasons = rejected_data['Judge Remarks'].value_counts().head(10)

rejection_reasons

Judge Remarks
APPLICANT WORKING AS A SUPERVISOR FOR 2 YEARS IN TFP RETAIL SDN BHD. APPLICANT HAS THE CAPACITY OF PAYING THE INSTALMENT AND THE DSR REFLECTED ONLY 21.13%. RECOMMENDED FOR APPROVAL._x000D_\nLOAN APPROVED AT RM29K. SUBJECT TO UTILITY BILL PROVIDED._x000D_\nCHANGE CAR FROM SYE9303 TO VEM6184. SAME CAR MODEL, SAME YEAR. OTHERS REMAIN UNCHANGE._x000D_\nIncorrect DSR. Pls amend_x000D_\nAPPLICANT WORKING AS A SUPERVISOR FOR 2 YEARS IN TFP RETAIL SDN BHD. APPLICANT HAS THE CAPACITY OF PAYING THE INSTALMENT AND THE DSR REFLECTED ONLY 21.13%. RECOMMENDED FOR APPROVAL._x000D_\nSales admin to input remarks_x000D_\nDONE AMENDMENT._x000D_\nProceed to approve subject to utility bill._x000D_\nCHANGE CAR FROM SYE9303 TO VEM6184. SAME CAR MODEL, SAME YEAR. OTHERS REMAIN UNCHANGE. _x000D_\n_x000D_\nCHANGE CAR MODEL FROM PERODUA BEZZA VEM6184 TO PROTON SAGA QS6971M , OTR PRICE CHANGE TO RM40,000 AND LOAN AMOUNT TO RM31,400. OTHER TERM REMAIN UNCHANGE._x000D_\ntrade ref<1.5k, CCRIS prompt co

In [28]:
# Analyze loan amounts for disbursed vs rejected/canceled applications
loan_amount_by_status = df2.groupby('Judge Status')['Loan Type'].count()

# Exploring the average loan tenure for disbursed vs rejected/canceled cases
avg_tenure_by_status = df2.groupby('Judge Status')['H Yrs Svr'].mean()

loan_amount_by_status, avg_tenure_by_status

(Judge Status
 CANCEL       237
 DISBURSED     16
 Name: Loan Type, dtype: int64,
 Judge Status
 CANCEL       2.264135
 DISBURSED    3.968750
 Name: H Yrs Svr, dtype: float64)

## Scen 3 outcome
In Genie dataset 2, a large majority of the applications were "CANCEL" indicating a high rejection rate. The company's underwriting process is either too stringent, or attracting many applicants that fail to meet the profile criteria. There may be missed opportunities for applicants who could apply if pre-screening was more effective.
To improve from this situation, stricter pre-screening proccesses must apply to ensure applicants meet baseline criteria. 

Common reasons for rejection include incomplete customer details and pending bank statements and payslips. this suggests that there is a lack of pre-application screenings and application follow-ups to insert key documents. 
A stricter documentation submission policy could be put in place to avoid the scenerio where applicants are lacking information to what documents they need to submit. 
Introducing automated checks/follow ups durring application for incomplete documents would reduce the need for cancellation while still managing risks. 

Under the "H Yrs Svr" column, we can see that applicants that were disbursed loans have an average of 3.97 years in service which was much higher than those that had their application cancelled (2.26 years). Prioritizing job stability contributes to migating default risk.